### DISCUSSION TITLE
Git URL Rewriting can break ability to "Browse"

### DISCUSSION THREAD

**Discussion by chrisc-ironnet** [magit/forge#436] [Support] [answered]
In order to use multiple GitHub accounts, I use Git URL-rewriting (sort of like [this example in the Tips and Tricks](https://github.com/magit/forge/wiki/Tips-and-Tricks#using-git-url-rewriting-instead) document, but crucially, the URL in my rewriting rules isn't `github.com`).

For example, one rewriting rule looks like this:

```gitconfig
[url "git@github-home"]
    insteadOf = git@github.com
```

Now in Emacs, when I do a `' b P` (which runs `forge-browse-pullreqs`), it opens an invalid url in the browser: https://github-home/cfclrk/py-demo/pulls.

Or in Elisp:

```emacs-lisp
(let ((default-directory "/Users/chris.clark/Projects/py-demo"))
  (forge--format (forge-get-repository 'stub) 'pullreqs-url-format))
```
```
"https://github-home/cfclrk/py-demo/pulls"
```

I feel like there's probably a way to tell `forge` to use `https://github.com` instead of `https://github-home` when doing a `forge-browse-*` command. Buuut, I haven't figured it out.

Anyone else have this problem or know how to make `forge` use `https://github.com` when you have rewriting rules like this?

**Comment by cfclrk** [comment]
Ugh, I used the wrong GitHub account to post that question... the tribulations of multiple GitHub accounts!

I was able to temporarily get around this locally by making forge use `git config --get remote.origin.url` instead of using `git remote get-url origin`.

```
---
 lisp/forge-repo.el | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lisp/forge-repo.el b/lisp/forge-repo.el
index 2810421..0ca755e 100644
--- a/lisp/forge-repo.el
+++ b/lisp/forge-repo.el
@@ -158,7 +158,7 @@ repository, if any."
                                (car (member (forge--get-remote) remotes))
                              (car remotes)))))
           (if-let ((url (and remote
-                             (magit-git-string "remote" "get-url" remote))))
+                             (magit-git-string "config" "--get" "remote.origin.url"))))
               (when-let ((repo (forge-get-repository url remote demand)))
                 (oset repo worktree (magit-toplevel))
                 repo)
-- 
```

Not sure if that's a good idea for others or not.

**Reply by tarsius** [reply]
I don't understand why you do this:

```gitconfig
[url "git@github-home"]
    insteadOf = git@github.com
```

the example you mention does the opposite

```gitconfig
[url "git@github.com:"]
	insteadOf = gh:
```

I think you need:

```gitconfig
[url "git@github.com:"]
	insteadOf = gh-home:
[url "git@github.com:"]
	insteadOf = gh-work:
```

```
Host gh-home
     HostName github.com
     User home-user
Host gh-work
     HostName github.com
     User work-user
```

**Reply by cfclrk** [reply]
Thanks so much for the response! 

I'm trying out a workflow like you mention; I think I'll probably end up sticking with it.

The motivation for my original rewriting rules is for git to automatically use the correct SSH key (...usually) without having to manually modify any git repo URLs. For example, given the following rewriting rules:

```gitconfig
[url "git@github-home"]
    insteadOf = git@github.com
[url "git@github-work:MyWorkOrg"]
    insteadOf = git@github.com:MyWorkOrg
```

and an SSH config:

```
Host github-home
    HostName       github.com
    User           git
    IdentityFile   ~/.ssh/github-home
    IdentitiesOnly yes

Host github-work
    HostName       github.com
    User           git
    IdentityFile   ~/.ssh/github-work
    IdentitiesOnly yes
```

I can `git clone git@github.com:MyWorkOrg/my-repo.git`, and it automatically uses the SSH key for my work account. (There's a little more to it, which I badly document [here][github-identities] -- it's a WIP).

There are a couple drawbacks, like the fact that the "work" SSH key is _only_ used for repos with URLs like `git@github.com:MyWorkOrg` - what if I want to work on some open source project using my work account?

I'm certainly no git expert; I only got this idea from reading blogs and whatnot. I thought this was how everyone was doing it.

[github-identities]: https://www.cfclrk.com/articles/github_identities.html

**Reply by cfclrk** [reply]
The workflow @tarsius mentions worked nicely.

Even so, I've decided to go back to my original rewriting rules because:

1. A lot of scripts out there hard-code "github.com" URLs (which is reasonable). In order for those to work, I need a way of distinguishing which authentication method to use for which repo URLs.
2. It's nice not to have to manually edit every repo URL.

I'm actually surprised both these approaches to URL rewriting work. How does git know which URL to use? The rewritten form or the form that's actually in the `.git/config` file? I guess it examines both.

So anyway, for now, I'm continuing to use this local diff in `lisp/forge-repo.el`:

``` diff
@@ -158,7 +158,7 @@ repository, if any."
                                (car (member (forge--get-remote) remotes))
                              (car remotes)))))
           (if-let ((url (and remote
-                             (magit-git-string "remote" "get-url" remote))))
+                             (magit-git-string "config" "--get" "remote.origin.url"))))
               (when-let ((repo (forge-get-repository url remote demand)))
                 (oset repo worktree (magit-toplevel))
                 repo)
```

**Comment by cfclrk** [comment, marked as answer]
For posterity (I should have followed up a long time ago!), I no longer have this problem.

Turns out: URL rewriting is not necessary if you have multiple GitHub accounts. 

Instead, it's much easier to set a `core.sshCommand` for each account in your git config. I documented my current workflow here: https://www.cfclrk.com/articles/github_identities.html

**Reply by tarsius** [reply]
If you have the time, please consider adding a "Using includeIf instead" section to https://github.com/magit/forge/wiki/Tips-and-Tricks#using-git-url-rewriting-instead.
